packages/@aws-cdk/aws-ec2/lib/cfn-init-elements.ts

```diff
@@ -0,0 +1,330 @@
+    if (!this.packages[packageType]) {
+      this.packages[packageType] = [];
+    }
+    this.packages[packageType].push(key);
+  }
+
+  /**
+   * Render the restart handles for use in an InitService declaration
+   * @internal
+   */
+  public _renderRestartHandles(): any {
+    const nonEmpty = <A>(x: A[]) => x.length > 0 ? x : undefined;
+
+    return {
+      commands: nonEmpty(this.commands),
+      files: nonEmpty(this.files),
+      packages: Object.keys(this.packages).length > 0 ? this.packages : undefined,
+      sources: nonEmpty(this.sources),
+    };
+  }
+}
+
+/**
+ * Base class for all CloudFormation Init elements
+ */
+export abstract class InitElement {
+
+  /**
+   * Returns the init element type for this element.
+   */
+  public abstract readonly elementType: string;
+
+  /**
+   * Called when the Init config is being consumed. Renders the CloudFormation
+   * representation of this init element, and calculates any authentication
+   * properties needed, if any.
+   *
+   * @param options bind options for the element.
+   * @internal
+   */
+  public abstract _bind(options: InitBindOptions): InitElementConfig;
+
+}
+
+/**
+ * Options for InitCommand
+ */
+export interface InitCommandOptions {
+  /**
+   * Identifier key for this command
+   *
+   * Commands are executed in lexicographical order of their key names.
+   *
+   * @default - Automatically generated
+   */
+  readonly key?: string;
+
+  /**
+   * Sets environment variables for the command.
+   *
+   * This property overwrites, rather than appends, the existing environment.
+   *
+   * @default - Use current environment
+   */
+  readonly env?: Record<string, string>;
+
+  /**
+   * The working directory
+   *
+   * @default - Use default working directory
+   */
+  readonly cwd?: string;
+
+  /**
+   * Command to determine whether this command should be run
+   *
+   * If the test passes (exits with error code of 0), the command is run.
+   *
+   * @default - Always run this command
+   */
+  readonly test?: string;
+
+  /**
+   * Continue running if this command fails
+   *
+   * @default false
+   */
+  readonly ignoreErrors?: boolean;
+
+  /**
+   * The duration to wait after a command has finished in case the command causes a reboot.
+   *
+   * Set this value to `InitCommandWaitDuration.none()` if you do not want to wait for every command;
+   * `InitCommandWaitDuration.forever()` directs cfn-init to exit and resume only after the reboot is complete.
+   *
+   * For Windows systems only.
+   *
+   * @default - 60 seconds
+   */
+  readonly waitAfterCompletion?: InitCommandWaitDuration;
+
+  /**
+   * Restart the given service(s) after this command has run
+   *
+   * @default - Do not restart any service
+   */
+  readonly serviceRestartHandles?: InitServiceRestartHandle[];
+}
+
+/**
+ * Represents a duration to wait after a command has finished, in case of a reboot (Windows only).
+ */
+export abstract class InitCommandWaitDuration {
```

Comment:
This type feels heavy handed to me, though I could be persuaded to keep it in. I will never like having to type so much though.